components/brave_wallet/common/fil_address.cc

```diff
@@ -0,0 +1,180 @@
+
+#include "base/logging.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "brave/third_party/argon2/src/src/blake2/blake2.h"
+#include "components/base32/base32.h"
+
+namespace brave_wallet {
+
+namespace {
+
+constexpr int kChecksumSize = 4;
+constexpr size_t kPublicKeySizeSecp256K = 20;
+constexpr size_t kAddressSizeSecp256K = 41;
+constexpr size_t kPublicKeySizeBLS = 48;
+constexpr size_t kAddressSizeBLS = 86;
+
+absl::optional<std::vector<uint8_t>> BlakeHash(
+    const std::vector<uint8_t>& payload,
+    size_t length) {
+  blake2b_state blakeState;
+  if (blake2b_init(&blakeState, length) != 0) {
+    VLOG(0) << __func__ << ": blake2b_init failed";
+    return absl::nullopt;
+  }
+  if (blake2b_update(&blakeState, payload.data(), payload.size()) != 0) {
+    VLOG(0) << __func__ << ": blake2b_update failed";
+    return absl::nullopt;
+  }
+  std::vector<uint8_t> result(length, 0);
+  if (blake2b_final(&blakeState, result.data(), length) != 0) {
+    VLOG(0) << __func__ << ": blake2b_final failed";
+    return absl::nullopt;
+  }
+  return result;
+}
+
+bool IsValidNetwork(const std::string& network) {
+  return network == mojom::kFilecoinTestnet ||
+         network == mojom::kFilecoinMainnet;
+}
+
+absl::optional<mojom::FilecoinAddressProtocol> ToProtocol(char input) {
+  if ((input - '0') == static_cast<int>(mojom::FilecoinAddressProtocol::BLS))
+    return mojom::FilecoinAddressProtocol::BLS;
+  if ((input - '0') ==
+      static_cast<int>(mojom::FilecoinAddressProtocol::SECP256K1))
+    return mojom::FilecoinAddressProtocol::SECP256K1;
+  return absl::nullopt;
+}
+
+}  // namespace
+
+FilAddress::FilAddress(const std::vector<uint8_t>& bytes,
+                       mojom::FilecoinAddressProtocol protocol,
+                       const std::string& network)
+    : protocol_(protocol), network_(network), bytes_(bytes) {
+  DCHECK(IsValidNetwork(network));
+}
+FilAddress::FilAddress() = default;
+FilAddress::FilAddress(const FilAddress& other) = default;
+FilAddress::~FilAddress() = default;
+
+bool FilAddress::IsEqual(const FilAddress& other) const {
+  return bytes_.size() == other.bytes_.size() &&
+         std::equal(bytes_.begin(), bytes_.end(), other.bytes_.begin()) &&
+         protocol_ == other.protocol_ && network_ == other.network_;
+}
+
+bool FilAddress::operator==(const FilAddress& other) const {
+  return IsEqual(other);
+}
+
+bool FilAddress::operator!=(const FilAddress& other) const {
+  return !IsEqual(other);
+}
+
+// Decodes Filecoin BLS/SECP256K addresses within rules
+// https://spec.filecoin.io/appendix/address/#section-appendix.address.string
+// |------------|----------|---------|----------|
+// |  network   | protocol | payload | checksum |
+// |------------|----------|---------|----------|
+// | 'f' or 't' |  1 byte  | n bytes | 4 bytes  |
```

Comment:
Suggest putting the payload format for BLS and Secp256K1 here in the comment.

Reply:
done